### PR TITLE
Enterprise Two Factor Authentication Report

### PR DESCRIPTION
**Pull request by jingcheng16**
## Product Description

This is the PR for Two Factor Authentication Report. This tile will live in Security Center Page.

**UI:**
image

**The downloaded report**
| Domain not having 2FA enforced |
|--------|
| enterpriseconsoletest |

**The API**
```shell
curl -s -X GET "http://localhost:8000/a/enterpriseconsoletest/enterprise/api/v1/twofactorauth/" -H "Authorization: ApiKey xxx:xxxxxxxx"

{"@odata.context": "http://localhost:8000/a/enterpriseconsoletest/enterprise/api/v1/twofactorauth/schema/#feed", "@odata.count": 1, "value": [{"domain_not_having_2fa_enfoced": "enterpriseconsoletest"}]}%
``` 

## Technical Summary

Ticket: https://dimagi.atlassian.net/browse/SAAS-16310

If a domain has 2fa enforced, `Domain.two_factor_auth` will be `True`, otherwise `False`

## Feature Flag

The whole page is hidden behind `enterprise_dashboard_improvements`. The FF will be removed after this PR is merged.

### Safety story

Verified locally and on staging

### QA Plan

Do not planning on QA

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/enterprise/enterprise.py [hunk 1]
```diff
@@ -561,3 +564,22 @@ def rows_for_domain(self, domain_obj):
         num_errors = sum([result['direction_count'] for result in results if result['error']])
 
         return [(domain_obj.name, num_sent, num_received, num_errors), ]
+
+
+class Enterprise2FAReport(EnterpriseReport):
+    title = gettext_lazy('Two Factor Authentication')
+    total_description = gettext_lazy('# of Domains not having 2FA enforced')
```
Comment (orangejenny) [inline corehq/apps/enterprise/enterprise.py:571]: User-facing text typically uses "project spaces" instead of "domains".

Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: Thank you! 8b0ce9411e65f8c5f210035238cb0e63f381b582

File: corehq/apps/enterprise/api/resources.py [hunk 2]
```diff
@@ -408,6 +408,19 @@ def get_primary_keys(self):
         return ('form_id', 'submitted',)
 
 
+class TwoFactorAuthResource(ODataEnterpriseReportResource):
+    domain_not_having_2fa_enforced = fields.CharField()
```
Comment (mjriley) [inline corehq/apps/enterprise/api/resources.py:412]: maybe switch `domain_without_2fa`?

File: corehq/apps/enterprise/enterprise.py [hunk 3]
```diff
@@ -568,11 +568,11 @@ def rows_for_domain(self, domain_obj):
 
 class Enterprise2FAReport(EnterpriseReport):
     title = gettext_lazy('Two Factor Authentication')
-    total_description = gettext_lazy('# of Domains not having 2FA enforced')
+    total_description = gettext_lazy('# of Project Spaces not having 2FA enforced')
```
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:571]: Probably change to "# of Projects Spaces without (or lacking?) 2FA"

**Review by mjriley — COMMENTED** [review]
Once the user-facing wording is changed, the PR looks good to approve

File: corehq/apps/enterprise/api/resources.py [hunk 4]
```diff
@@ -408,6 +408,19 @@ def get_primary_keys(self):
         return ('form_id', 'submitted',)
 
 
+class TwoFactorAuthResource(ODataEnterpriseReportResource):
+    domain_without_2fa = fields.CharField()
```
Comment (mjriley) [inline corehq/apps/enterprise/api/resources.py:412]: Looking directly below, at the `CommCareVersionComplianceResource`, it does seem like we use `project_space` for any user-facing things,so `domain_without_2fa` should be changed to `project_space_without_2fa`?

Comment (jingcheng16) [reply corehq/apps/enterprise/api/resources.py:412]: `CommCareVersionComplianceResource` might be an exception, all other resources are using `domain`... Maybe it is `CommCareVersionComplianceResource` need to use `domain` ?

Comment (mjriley) [reply corehq/apps/enterprise/api/resources.py:412]: Yep, good catch.  We should probably update `CommCareVersionComplianceResource`, but it looks like this one is good to go

**Review by mjriley — APPROVED** [review]

**Pull request merged**